extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java

```diff
@@ -104,20 +104,17 @@ protected IcebergTableLocationKey locationKey(
 
     /**
      * @param tableAdapter The {@link IcebergTableAdapter} that will be used to access the table.
-     * @param tableSnapshot The {@link Snapshot} from which to discover data files.
      * @param instructions The instructions for customizations while reading.
      */
     public IcebergBaseLayout(
             @NotNull final IcebergTableAdapter tableAdapter,
-            @Nullable final Snapshot tableSnapshot,
-            @NotNull final IcebergInstructions instructions,
+            @NotNull final IcebergReadInstructions instructions,
             @NotNull final DataInstructionsProviderLoader dataInstructionsProvider) {
         this.tableAdapter = tableAdapter;
-        this.snapshot = tableSnapshot;
+        this.snapshot = instructions.snapshot().orElse(null);
```

Comment:
Is this incorrect now? Do we need to take into account snapshotId potentially? I'm assuming in almost all cases internally, we need to call something like `tableAdapter.getSnapshot` instead?

Reply:
This is a public method though. At a minimum, we need to document this assumption in javadoc and should be defensive here: if `.snapshot()` is not present, then error if `.snapshotId()` is present (this could even be elevated to a method on `IcebergReadInstructions` if we find ourselves relying on it). Or, the code should be refactored that this constructor can only be called internally (and even in that case, we should still mandate a safety check). 

I'm worried about this implicit assumptions that we don't make explicit. Very much in the same vein as https://github.com/deephaven/deephaven-core/pull/6149.